### ISSUE TITLE
Do not validate checkpointDate online if scanForPackages = FALSE

### ISSUE THREAD

**Issue by ebs238** [RevolutionAnalytics/checkpoint#234]
new to checkpoint, so probably missing something here, but...

i am completely unable to run any script because the MRAN servers are down. i have tried setting `scanForPackages=F`, but still not luck. 

packages in the chekpointlocation are available and valid. i like this versioning scheme, but did not realize i was getting into a real-time critical dependency on and external server

**Comment by j-martens** [comment]
Thanks for your report! It should be back up soon.

**Comment by ebs238** [comment]
I appreciate that they are back up, but i was trying to point what i think is an issue with this library: it makes a project entirely unusable when the servers are down. I think that's a bit of an extreme dependency. It would be great if there was a switch/flag/parameter that allowed scripts to run against an existing cache in such scenarios.

**Comment by andrie** [comment]
The intention of `scanForPackages = FALSE` is to avoid precisely the situation you are describing.

If this doesn't work, then potentially you have found a bug.

Can you make a reproducible example of this, so I can take a look?

**Comment by ebs238** [comment]
Just did some testing and it looks like it does indeed work... it just takes a very long time. It seems that even with `scanForPackages = FALSE` there is still a dependency on the MRAN servers that has to time out. In my case, that timeout is taking over 3 minutes, which is what what led me to believe it was failing.

I'm not sure whether it tries to download each library and wait for each timeout, or if it just using a very large timeout value, but could probably save the next person some grief by reducing these waits.

![untitled](https://cloud.githubusercontent.com/assets/24379655/23769490/ca236c14-04cc-11e7-951e-18c717b34c18.png)

**Comment by andrie** [comment]
Thanks for this report - this is very helpful.

You are quite correct. When `scanForPackages = FALSE`, we still attempt to validate that the `checkpointDate` is valid, by pinging MRAN.

This doesn't make sense.

Fix: Do not check online for valid snapshot date if `scanForPackages = FALSE`.

**Comment by andrie** [comment]
Added a timout of 5 seconds in commit d9b269d65f8b69f300b1fe6b6e140e5b33de5a54

**Comment by andrie** [comment]
@ebs238 This change is now in the `master` branch on github. Can you please test and report if the behaviour is better?

**Comment by ebs238** [comment]
apologies, but I have no idea how to install from a branch. here is the process i used for my initial test if you want to try that:

edit host file to set mran to an unreachable address. I used the private 10.x space, but this will obviously depend on your network. you must use an unreachable address, not for this test, as localhost, as localhost will fail quickly
`10.0.0.1                mran.microsoft.com
10.0.0.1                 mran.revolutionanalytics.com`

I think i had to restart my IDE to pick up the change in the host file

**Comment by andrie** [comment]
To install from github, try:

```r
devtools::install_github("RevolutionAnalytics/checpoint")
```

**Comment by andrie** [comment]
To install from github, try:

```r
devtools::install_github("RevolutionAnalytics/checkpoint")
```

**Comment by ebs238** [comment]
Does not seem to be working
![untitled](https://cloud.githubusercontent.com/assets/24379655/24435772/91d04d64-13f4-11e7-9436-10f56a2e7bc6.png)

**Comment by andrie** [comment]
Thanks for your report.  This is very strange, and I'll have to investigate further.

I tried to replicate by setting the `mranUrl` to a non-existent URL:

```r
> options(checkpoint.mranUrl = "https://foobah")
> system.time({
+   checkpoint("2014-09-25", project = "C:/Users/adevries/Dropbox/checkpoint/checkpoint-demo")
+ })
```

Then the timeout works as expected, and the process takes 10 seconds:

```
Unable to connect to MRAN. Skipping some date validations.
Scanning for packages used in this project
  |===========================================================================| 100%
- Discovered 2 packages
All detected packages already installed
checkpoint process complete
---
   user  system elapsed 
   0.00    0.05   11.36 
```

**Comment by ebs238** [comment]
I just did a quick test and I match your results for an invalid `mranUrl`. However, invalid host file entries still takes multiple minutes to timeout (`checkpoint_0.3.18`)

**Comment by andrie** [comment]
Yes, these changes are not available in `master` yet.  However, the `dev` branch contains a substantial improvement - it doesn't make any online validation calls and completes the checkpoint process in fractions of a second.

**Comment by andrie** [comment]
The enhancements are now available in `v0.3.20` in the master branch on github.
